src/core/qgsvectorlayer.cpp

```diff
@@ -2243,10 +2248,34 @@ bool QgsVectorLayer::changeGeometry( QgsFeatureId fid, const QgsGeometry &geom )
 
 bool QgsVectorLayer::changeAttributeValue( QgsFeatureId fid, int field, const QVariant &newValue, const QVariant &oldValue )
 {
-  if ( !mEditBuffer || !mDataProvider )
-    return false;
+  if ( fields().fieldOrigin( field ) == QgsFields::OriginJoin )
+  {
+    int srcFieldIndex;
+    const QgsVectorLayerJoinInfo *info = mJoinBuffer->joinForFieldIndex( field, fields(), srcFieldIndex );
+    if ( info && info->joinLayer() && info->isEditable() )
+    {
+      QgsFeature feature = getFeature( fid );
+
+      if ( !feature.isValid() )
+        return false;
+
+      QgsFeature joinFeature = mJoinBuffer->joinedFeatureOf( info, feature );
 
-  return mEditBuffer->changeAttributeValue( fid, field, newValue, oldValue );
+      if ( joinFeature.isValid() )
+        return info->joinLayer()->changeAttributeValue( joinFeature.id(), srcFieldIndex, newValue, oldValue );
```

Comment:
I've asked myself the same question. But I think it would be too permissive/dangerous to automatically set the joined layer as editable. However, I don't have a strong opinion here.

@haubourg What do you think about it?